doc/developer/design/20230720_single_statement_explitic_transaction.md

```diff
@@ -0,0 +1,105 @@
+We will then describe how this design modifies that state machine but still meets all invariants.
+
+Each `Session` has a `tranasction` property of type `enum TransactionStatus`.
+It is the responsibility of the various protocol handlers (pgwire, http, ws) to correctly manage this.
+A `TransactionStatus` is an enum with variants:
+
+- `Default`
+- `Started`
+- `InTransaction`
+- `InTransactionImplicit`
+- `Failed`
+
+`Default` is the initial state, and also the state when there is no in-progress transaction.
+In order for a session to execute a statement, the protocol handler must set this to one of the other states then end the transaction with a commit or rollback.
+`Started` is used for a single statement, so an immediate and implied commit follows.
+`InTransaction` is used when a user explicitly types `BEGIN`, and can only be exited when a user sends `COMMIT` or `ROLLBACK`.
+`InTransactionImplicit` is used when a user sends multiple statements in the same query (`SELECT 1; SELECT 2`), and an implicit `BEGIN` and `COMMIT` wrap it.
+`Failed` can only be entered when a statement in a transaction has error'd from the `InTransaction` state.
+If failures occur in the other in-progress states, an implicit `ROLLBACK` must be sent by the handler.
+
+All status except `Default` have an inner `Transaction` object that tracks what operations have happened in a transaction determined by the first statement.
+These operations are: `Peeks`, `Subscribe`, `Writes`, `None`.
+`None` is the default and can transition to any other operation.
+The other operations can be merged with like operations, except `Subscribe` which supports only a single `SUBSCRIBE`.
+
+The `ReadyForQuery` message is sent by the server when it is ready to receive new queries, and it includes the current transaction status code: `Idle`, `InTransaction`, or `Failed`.
+`InTransaction` and `Failed` correspond to the similarly named variants above and can only appear in explicit transactions.
+`Idle` is return in the `Default` state, because all other transaction states implicitly close and should never be active when deciding the status code.
+
+The pgwire, http, and ws handlers enforce these requirements and state changes.
+
+#### Changes from this design
+
+This design adds a new operation to the inner transaction: `SingleStatement`.
+This operation records a single statement that is later retrieved and executed during `COMMIT`.
+
+This design also adds what is essentially a new handler: the coordinator itself.
+When a transaction is being committed and its inner operation is `SingleStatement`, the inner statement is queued in the coordinator for execution by a new method that is serving as a handler: `sequence_execute_single_statement_transaction`.
+This method:
+1. Ensures the session's transaction is in `Default`, which should be a side effect of running it through `sequence_end_transaction` from the `COMMIT`.
```

Comment:
I was finding this a bit confusing as I read through the code, it seems like the status really should be in the `InTransaction` state because it's in an explicit transaction that hasn't committed yet. The fact that this transaction goes through two cycles of default->in-transaction->commit is a bit hard to follow and reason about.

Reply:
I agree it's weird because of the extra cycle. This feature is essentially a new transaction operation: `SingleStatement`. The other operations (`Peeks`, "Writes`) already do things *after* the transaction has cleared but during "operation processing" time. Peeks can ship off to the strict serializable wait, Writes can ship off to PendingWrites. `SingleStatement`'s operation processing is to enqueue a new coord message that executes the recorded statement. I've added some text to explain more.